src/IECoreMaya/SceneShapeSubSceneOverride.cpp

```diff
@@ -237,378 +185,415 @@ class Expandulator
-	case IECoreScene::TypeId::PointsPrimitiveTypeId :
-		fillPointsData( key.object.get(), geometryData.get() ); break;
-	case IECoreScene::TypeId::CurvesPrimitiveTypeId :
-		fillCurvesData( key.object.get(), geometryData.get() ); break;
-	default :
-		// we should never get here as only the above types are supported, see `objectCanBeRendered`.
-		break;
+		msg( Msg::Error, "SceneShapeSubSceneOverride::bufferGetter", ( boost::format( "Memory acquisition for %s buffer failed." ) % descriptor.semanticName() ).str() );
 	}
 
-	cost = computeGeometryDataCost( geometryData.get() );
-
-	return geometryData;
+	return buffer;
 }
 
-using GeometryDataCache = IECore::LRUCache<IECore::MurmurHash, GeometryDataPtr, IECore::LRUCachePolicy::Parallel, GeometryDataCacheGetterKey>;
-
-GeometryDataCache &geometryDataCache()
+VertexBufferPtr createVertexBuffer( const MVertexBufferDescriptor &descriptor, const V2fVectorData *data )
 {
-	static GeometryDataCache g_geometryDataCache( geometryGetter, memoryLimit() * 0.25 );
-	return g_geometryDataCache;
+	VertexBufferPtr buffer( new MVertexBuffer( descriptor ) );
+
+	size_t numEntries = data->readable().size();
+	void *bufferData = buffer->acquire( numEntries, true );
+	if( bufferData && buffer )
+	{
+		memcpy( bufferData, data->baseReadable(), sizeof( float ) * 2 * numEntries );
+		buffer->commit( bufferData );
+	}
+	else
+	{
+		msg( Msg::Error, "SceneShapeSubSceneOverride::bufferGetter", "Memory acquisition for uv buffer failed." );
+	}
+
+	return buffer;
 }
 
-////////////////////////////////////////////////////////////////////////////////////////////
-// Maya Buffer Cache
-////////////////////////////////////////////////////////////////////////////////////////////
+using CacheEntry = boost::variant<VertexBufferPtr, ExpandulatorPtr>;
+using CacheEntryPtr = std::shared_ptr<CacheEntry>;
 
 boost::signals2::signal<void (const BufferPtr & )> &bufferEvictedSignal()
 {
 	static boost::signals2::signal<void (const BufferPtr & )> g_bufferEvictedSignal;
 	return g_bufferEvictedSignal;
 }
 
-struct BufferCleanup
+struct CacheCleanup
 {
-	void operator() ( const IECore::MurmurHash &key, const BufferPtr &buffer )
+	void operator() ( const IECore::MurmurHash &key, const CacheEntryPtr &entry )
 	{
-		bufferEvictedSignal()( buffer );
+		if( auto buffer = boost::get<VertexBufferPtr>( entry.get() ) )
+		{
+			bufferEvictedSignal()( std::make_shared<Buffer>( *buffer ) );
+		}
+		else if( auto expandulator = boost::get<ExpandulatorPtr>( entry.get() ) )
+		{
+			bufferEvictedSignal()( std::make_shared<Buffer>( expandulator->get()->indices() ) );
+
+			auto wireframeIndices = expandulator->get()->wireframeIndices( /* force */ false );
+			if( wireframeIndices && wireframeIndices != expandulator->get()->indices() )
+			{
+				bufferEvictedSignal()( std::make_shared<Buffer>( wireframeIndices ) );
+			}
+		}
 	}
 };
 
-struct BufferCacheGetterKey
+struct CacheKey
```

Comment:
Its considerably less expensive than the combined previous keys. We use to hash the topology to get the `Expandulator`, hash the entire primitive (including primvars we don't read from) to get the `GeometryData` struct, and finally hash the primvar we're after. Now we hash the topology and the primvar we're after.

> I wonder how often you would actually lose out on caching something if you just used the addresses of the buffers rather than actually hashing all the data.

I'm not sure, but I wouldn't want to introduce a change like that without first profiling to see the hash as a bottleneck, and doing some thorough testing on a set if representative production scenes.

Reply:
That said, I think I can be opting out of some of this, depending if an `Expandulator` is being requested or a `VertexBuffer`... will do some testing.

Reply:
Hashing improved in a71f1e7 (though I'm still hashing the actual data for the reasons stated above).